ulib/FStar.FunctionalExtensionality.fst

```diff
@@ -19,28 +19,70 @@ module FStar.FunctionalExtensionality
 
 inline_for_extraction
 let on_domain (a:Type) (#b:a -> Type) (f:arrow a b)
-  = f
+  = fun (x:a) -> f x
 
 let feq_on_domain (#a:Type) (#b:a -> Type) (f:arrow a b)
   = ()
 
 let idempotence_on_domain #a #b f
-  = ()
+  = assert_norm (on_domain a f == (on_domain a (on_domain a f)))
+
+let quantifier_as_lemma (#a:Type) (#b: a -> Type)
+                        (f:squash (forall (x:a). b x))
+                        (x:a)
+    : Lemma (b x)
+    = ()
+
+let extensionality_1 (a:Type)
+                     (b: a -> Type)
+                     (f g: arrow a b)
+                     (sq_feq : squash (feq f g))
+  : Lemma (ensures on_domain a f == on_domain a g)
+  = admit()
+    // let open FStar.Tactics in
```

Comment:
These are actually provable with tactics.

But, using tactics in funext leads to circularities in bootstrapping the compiler.

A separate PR will fix those bootstrapping problems and remove these admits